src/main/java/org/springframework/integration/aws/outbound/S3MessageHandler.java

```diff
@@ -312,40 +309,58 @@ private Transfer upload(Message<?> requestMessage) {
 			if (this.uploadMetadataProvider != null) {
 				this.uploadMetadataProvider.populateMetadata(metadata, requestMessage);
 			}
-			InputStream inputStream;
-			if (payload instanceof InputStream) {
-				inputStream = (InputStream) payload;
-			}
-			else if (payload instanceof File) {
-				File fileToUpload = (File) payload;
-				if (key == null) {
-					key = fileToUpload.getName();
-				}
-				try {
-					inputStream = new FileInputStream(fileToUpload);
 
+			PutObjectRequest putObjectRequest = null;
+
+			try {
+				if (payload instanceof InputStream) {
+					InputStream inputStream = (InputStream) payload;
+					Assert.state(inputStream.markSupported(),
+							"The markSupported() method for an InputStream must evaluate to " +
+							"true for an upload request. ");
+					String contentMd5 = Md5Utils.md5AsBase64(inputStream);
```

Comment:
The AWS docs state that not specifying Content-Length for an InputStream will incur a significant performance hit.  Given that we are incurring an equivalent performance hit computing the hash value here, we would ideally compute Content-Length also.